### PR TITLE
Add the SMT solver Bitwuzla to JavaSMT

### PR DESCRIPTION
**Pull request by baierd**
This PR discusses the addition of Bitwuzla to JavaSMT.
This PR is based on 2 previous PRs:
  - [PR332](https://github.com/sosy-lab/java-smt/pull/332) is the original PR adding Bitwuzla support to JavaSMT

  - [PR369](https://github.com/sosy-lab/java-smt/pull/369) is an addition that combines PR332 with an automatic, SWIG based build script for the JNI wrapper and the most recent update of Bitwuzla with its new term manager, allowing solver instance and thread independent creation/handling of terms. The automatic, SWIG based build script generates a new wrapper based on the C++ bindings automatically for Bitwuzla. This can be toggled on and off. Also, this handles resources automatically via phantom references. 

Remaining problems:
- There is no support to translate FPs to decimal numbers. We are currently working on a general and independent Java based solution for all solvers and more for this.
- We got FILE handle independent parsing support in the most recent Bitwuzla update. While we integrated this, parsing now is kinda a mess and we should ask for a function that translates a full SMTLIB2 file into a single Term.

### CODE REVIEW

File: src/org/sosy_lab/java_smt/test/NonLinearArithmeticWithModuloTest.java [hunk 1]
```diff
@@ -34,7 +34,8 @@ public class NonLinearArithmeticWithModuloTest extends SolverBasedTest0 {
   @Parameters(name = "{0} {1}")
   public static Iterable<Object[]> getAllSolversAndTheories() {
     return Lists.cartesianProduct(
-            Arrays.asList(Solvers.values()), Arrays.asList(NonLinearArithmetic.values()))
```
Comment (baierd) [inline src/org/sosy_lab/java_smt/test/NonLinearArithmeticWithModuloTest.java:37]: Please revert.

**Comment by baierd** [comment]
@kfriedberger there are some additions in the tests that you should look at and report back if you find something that you don't like.


General note: the TermManager can NOT be initialized once and reused in multiple contexts!

**Review by baierd — COMMENTED** [review]
@daniel-raffler i found some minor things that we should take care of before merging.

Note: the review is missing Model and Creator.

File: src/org/sosy_lab/java_smt/test/FloatingPointFormulaManagerTest.java [hunk 2]
```diff
@@ -831,21 +833,27 @@ private List<Float> getListOfFloats() {
...
+
+    final int stepSize = solverToUse() == Solvers.BITWUZLA ? 10 : 1;
+    for (int i = 1; i < 20; i += stepSize) {
+      for (int j = 1; j < 20; j += stepSize) {
         flts.add((float) (i * Math.pow(10, j)));
         flts.add((float) (-i * Math.pow(10, j)));
       }
     }
 
+    final int numRandom = solverToUse() == Solvers.BITWUZLA ? 5 : NUM_RANDOM_TESTS;
```
[the remark this answers is not in the document]
Comment (baierd) [reply src/org/sosy_lab/java_smt/test/FloatingPointFormulaManagerTest.java:854]: Is this caused by any of our workarounds/additions?
If not, we should extract this into a native program and report this.
And also please comment the reasoning for this in the code.

[the remark this answers is not in the document]
Comment (daniel-raffler) [reply src/org/sosy_lab/java_smt/test/FloatingPointFormulaManagerTest.java:854]: I don't think it is caused by the workaround. `checkIeeeBv2FpConversion32` casts from bitvector to float and the workaround is only needed for the opposite direction. According to the profiler we're spending almost all our time on the `assertEqualsAsFp` on line 910, which then calls `Bitwuzla.check_sat` from the native bindings (and also spends another 8 seconds on closing the solver instances). Maybe the performance could be improved by re-using the ProverEnvironment?

File: src/org/sosy_lab/java_smt/solvers/bitwuzla/BitwuzlaNativeApiTest.java [hunk 3]
```diff
@@ -0,0 +1,622 @@
...
+    assertThat(res).isEqualTo(Result.SAT);
+
+    assertThat(xExt.num_children()).isEqualTo(1);
+    assertThat(xExt.num_indices()).isGreaterThan(0);
+    Vector_Int indices = xExt.indices();
+    assertThat(indices.size()).isEqualTo(1);
+    assertThat(indices.get(0)).isEqualTo(2);
+  }
+
+  // Todo:
```
[the remark this answers is not in the document]
Comment (daniel-raffler) [reply src/org/sosy_lab/java_smt/solvers/bitwuzla/BitwuzlaNativeApiTest.java:517]: Do you remember what this test was meant to demonstrate? If I run it the result is SAT, which seems fine to me. Then again it uses `forall` instead of `exists`?

File: lib/native/source/libbitwuzla/src/org/sosy_lab/java_smt/solvers/bitwuzla/api/Bitwuzla.java [hunk 4]
```diff
@@ -0,0 +1,99 @@
+/* ----------------------------------------------------------------------------
```
[the remark this answers is not in the document]
Comment (daniel-raffler) [reply lib/native/source/libbitwuzla/src/org/sosy_lab/java_smt/solvers/bitwuzla/api/Bitwuzla.java:9]: I've added license headers to all the auto-generated files. As far I know there is no way to do this automatically in SWIG and the headers will get overwritten when the bindings are updated. We could however write our own script to patch them back in.

File: src/org/sosy_lab/java_smt/solvers/bitwuzla/BitwuzlaNativeApiTest.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (baierd) [reply src/org/sosy_lab/java_smt/solvers/bitwuzla/BitwuzlaNativeApiTest.java]: I think i was just lazy and copied the `forall` test. I fixed it now and added model checking. Thanks for the reminder.

File: src/org/sosy_lab/java_smt/test/FloatingPointFormulaManagerTest.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (baierd) [reply src/org/sosy_lab/java_smt/test/FloatingPointFormulaManagerTest.java:854]: We should open an issue for this, extract a problem and submit a issue to Bitwuzla. I've opened issue #371 for this.

File: build/build-publish-solvers/solver-bitwuzla.xml [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+This file is part of JavaSMT,
+an API wrapper for a collection of SMT solvers:
+https://github.com/sosy-lab/java-smt
+
+SPDX-FileCopyrightText: 2024 Dirk Beyer <https://www.sosy-lab.org>
+
+SPDX-License-Identifier: Apache-2.0
+-->
+
+<!-- vim: set tabstop=8 shiftwidth=4 expandtab sts=4 filetype=ant fdm=marker: -->
```
[the remark this answers is not in the document]
Comment (daniel-raffler) [reply build/build-publish-solvers/solver-bitwuzla.xml:13]: I've now made it mandatory to specify if the wrapper should be rebuild and added a proper error message if the option is missing. I can add some more documentation to the code later this evening.

GMP is not actually build by the script and needs to be compiled by the user first. I could, however, still include this as another step?

[the remark this answers is not in the document]
Comment (baierd) [reply build/build-publish-solvers/solver-bitwuzla.xml:13]: You can add it with another flag and a error msg if the correct libs are not found. But this is not time sensitive and can be done later.

File: build/build-publish-solvers/solver-bitwuzla.xml [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+        </exec>
+
+        <!-- patch JNI library as it fails to do later for some reason -->
+        <exec executable="patchelf" failonerror="true">
+            <arg value="--set-rpath"/>
+            <arg value="$ORIGIN"/>
+            <arg value="--replace-needed"/><arg value="libbitwuzla.so.0"/><arg value="libbitwuzla.so"/>
+            <arg value="${source.path}/libbitwuzlaJNI.so"/>
+        </exec>
+        <exec executable="patchelf" failonerror="true">
```
[the remark this answers is not in the document]
Comment (kfriedberger) [reply build/build-publish-solvers/solver-bitwuzla.xml:167]: I have no concrete idea about it, just a guess: it is possible that there are multiple entries in the library and patchelf only changes the first one.

File: src/org/sosy_lab/java_smt/solvers/bitwuzla/BitwuzlaFormulaManager.java [hunk 7]
```diff
@@ -0,0 +1,265 @@
...
+        out.append(dump);
+      }
+    };
+  }
+
+  static Term getBitwuzlaTerm(Formula pT) {
+    return ((BitwuzlaFormula) pT).getTerm();
+  }
+
+  // Split up a sequence of lisp expressions
```
Comment (kfriedberger) [inline src/org/sosy_lab/java_smt/solvers/bitwuzla/BitwuzlaFormulaManager.java:209]: Are we sure we want to do tokenization for SMTLIB on our own?

Does this consider special cases like `(define-const |(some weird name)| Int)` and string constants? :-)

File: src/org/sosy_lab/java_smt/solvers/bitwuzla/BitwuzlaModel.java [hunk 8]
```diff
@@ -0,0 +1,238 @@
...
+      return name;
+    }
+    return getArrayName(array.get(0));
+  }
+
+  private Term buildEqForTwoTerms(Term left, Term right) {
+    assert left.sort().equals(right.sort());
+    Kind kind = Kind.EQUAL;
+    if (left.sort().is_fp()) {
+      kind = Kind.FP_EQUAL;
```
Comment (kfriedberger) [inline src/org/sosy_lab/java_smt/solvers/bitwuzla/BitwuzlaModel.java:205]: just asking: for FP theory, the operation Kind.FP_Equal does handle NaN-equalities, right?

File: src/org/sosy_lab/java_smt/solvers/bitwuzla/BitwuzlaQuantifiedFormulaManager.java [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+      substVars[i] = boundCopy;
+    }
+
+    Map_TermTerm map = new Map_TermTerm();
+    for (int i = 0; i < origVars.length; i++) {
+      map.put(origVars[i], substVars[i]);
+    }
+    body = termManager.substitute_term(body, map);
+
+    Term[] argsAndBody = new Term[2];
```
Comment (kfriedberger) [inline src/org/sosy_lab/java_smt/solvers/bitwuzla/BitwuzlaQuantifiedFormulaManager.java:60]: It might be simpler to understand if this 2-element-array is moved into the loop completely and we do not reuse it for several iterations.

File: src/org/sosy_lab/java_smt/solvers/bitwuzla/BitwuzlaSolverContext.java [hunk 10]
```diff
@@ -0,0 +1,266 @@
...
+            .withKeyValueSeparator(Splitter.on('=').limit(2).trimResults());
+    ImmutableMap<String, String> furtherOptionsMap;
+
+    try {
+      furtherOptionsMap = ImmutableMap.copyOf(optionSplitter.split(pFurtherOptions));
+    } catch (IllegalArgumentException e) {
+      throw new InvalidConfigurationException(
+          "Invalid Bitwuzla option in \"" + pFurtherOptions + "\": " + e.getMessage(), e);
+    }
+    for (Map.Entry<String, String> option : furtherOptionsMap.entrySet()) {
```
Comment (kfriedberger) [inline src/org/sosy_lab/java_smt/solvers/bitwuzla/BitwuzlaSolverContext.java:158]: Reflection-based setting of options is no clean solution.
The alternative would be a larger switch-case that sets each option by name.
Hmmm... It might be working, so we could keep it.

File: src/org/sosy_lab/java_smt/solvers/opensmt/OpenSmtFormulaManager.java [hunk 11]
```diff
@@ -83,4 +83,24 @@ public void appendTo(Appendable out) throws IOException {
       }
     };
   }
+
```
Comment (kfriedberger) [inline src/org/sosy_lab/java_smt/solvers/opensmt/OpenSmtFormulaManager.java:86]: This code change is not related to Bitwuzla and should be done in another PR.

Additionally, it might not work with escaped names and string constants in SMTLIB.

**Review by kfriedberger — COMMENTED** [review]
I just read the PR, I have not tested it.
The overall code looks good.
I added a few comments.

File: src/org/sosy_lab/java_smt/solvers/bitwuzla/BitwuzlaModel.java (hunk 8 above)
Comment (daniel-raffler) [reply src/org/sosy_lab/java_smt/solvers/bitwuzla/BitwuzlaModel.java:205]: It's poorly documented, but I think that Kind.FP_Equals sets -0=0 and nan/=nan, whereas it's -0/=0 and nan=nan for Kind.Equals.

File: lib/native/source/libbitwuzla/bitwuzla_wrap.cpp [hunk 12]
```diff
@@ -0,0 +1,7093 @@
+// This file is part of JavaSMT,
+// an API wrapper for a collection of SMT solvers:
+// https://github.com/sosy-lab/java-smt
+//
+// SPDX-FileCopyrightText: 2024 Dirk Beyer <https://www.sosy-lab.org>
+//
+// SPDX-License-Identifier: Apache-2.0
+
+/* ----------------------------------------------------------------------------
```
[the remark this answers is not in the document]
Comment (daniel-raffler) [reply lib/native/source/libbitwuzla/bitwuzla_wrap.cpp:9]: bitwuzla_wrap.cpp is generated in the build script [here](https://github.com/sosy-lab/java-smt/blob/5b7fb542fbe19d096974fe53457778951ea055b7/build/build-publish-solvers/solver-bitwuzla.xml#L126). The interface file ([bitwuzla.i](https://github.com/sosy-lab/java-smt/blob/5b7fb542fbe19d096974fe53457778951ea055b7/lib/native/source/libbitwuzla/bitwuzla.i)) is the input to SWIG and we get the Java proxies and the C++ wrapper code as output.

File: src/org/sosy_lab/java_smt/solvers/bitwuzla/BitwuzlaFloatingPointManager.java [hunk 13]
```diff
@@ -0,0 +1,355 @@
...
+    // FIXME: We should use a reserved symbol for the fresh variables
+    int sizeExp = pNumber.sort().fp_exp_size();
+    int sizeSig = pNumber.sort().fp_sig_size();
+
+    Sort bvSort = termManager.mk_bv_sort(sizeExp + sizeSig);
+
+    // Note that NaN is handled as a special case in this method. This is not strictly necessary,
+    // but if we just use "fpTerm = to_fp(bvVar)" the NaN will be given a random payload (and
+    // sign). Since NaN payloads are not preserved here anyway we might as well pick a canonical
+    // representation.
```
[the remark this answers is not in the document]
Comment (daniel-raffler) [reply src/org/sosy_lab/java_smt/solvers/bitwuzla/BitwuzlaFloatingPointManager.java:221]: I did ask the Bitwuzla developers about this method and [this](https://github.com/bitwuzla/bitwuzla/discussions/103#discussioncomment-8339056) is the response I got. Then again I do remember seeing one false positive when running CPAchecker on the sv-comp set with the workaround enabled. We could disable it for now, but this will break a lot of other programs from the sv-comp set that need to cast from float to bitvector.

File: lib/native/source/libbitwuzla/bitwuzla.i [hunk 14]
```diff
@@ -0,0 +1,328 @@
...
+%ignore Bitwuzla::print_formula (std::ostream &out, const std::string &format="smt2") const;
+%extend Bitwuzla {
+  std::string print_formula () {
+    std::ostringstream out;
+    $self->print_formula(out);
+    return out.str();
+  }
+}
+%ignore Bitwuzla::statistics () const;
+%ignore Bitwuzla::simplify ();
```
[the remark this answers is not in the document]
Comment (daniel-raffler) [reply lib/native/source/libbitwuzla/bitwuzla.i:303]: From the documentation: [simplify](https://bitwuzla.github.io/docs/cpp/classes/bitwuzla.html#_CPPv4N8bitwuzla8Bitwuzla8simplifyEv)
I'm not sure what it does exactly, but there is no reason not to include it.

File: src/org/sosy_lab/java_smt/solvers/bitwuzla/BitwuzlaQuantifiedFormulaManager.java (hunk 9 above)
Comment (daniel-raffler) [reply src/org/sosy_lab/java_smt/solvers/bitwuzla/BitwuzlaQuantifiedFormulaManager.java:60]: I think this code is essentially `foldr mkForall body [var1, var2,..]`. But there may be a simpler way to write it?

File: build/build-publish-solvers/solver-bitwuzla.xml (hunk 6 above)
[the remark this answers is not in the document]
Comment (baierd) [reply build/build-publish-solvers/solver-bitwuzla.xml:196]: ldd says that there is only one. @PhilippWendler maybe you have an idea?

File: build/build-publish-solvers/solver-bitwuzla.xml [hunk 15]
```diff
@@ -0,0 +1,236 @@
...
+            <arg value="org.sosy_lab.java_smt.solvers.bitwuzla.api"/>
+            <arg value="-outdir"/>
+            <arg value="src/org/sosy_lab/java_smt/solvers/bitwuzla/api"/>
+            <arg value="-o"/>
+            <arg value="bitwuzla_wrap.cpp"/>
+            <arg value="bitwuzla.i"/>
+        </exec>
+
+        <!-- apply patch for the phantom reference issue -->
+        <!-- FIXME: This will lead to memory leaks. We should look for a proper fix. -->
```
[the remark this answers is not in the document]
Comment (baierd) [reply build/build-publish-solvers/solver-bitwuzla.xml:131]: I don't think that this applies in the new version of Bitwuzla with the TermManager as we can destroy all associated terms/sorts of an TermManager when destroying it.

File: lib/native/source/libbitwuzla/src/org/sosy_lab/java_smt/solvers/bitwuzla/api/Result.java [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+      return swigValues[swigValue];
+    for (int i = 0; i < swigValues.length; i++)
+      if (swigValues[i].swigValue == swigValue)
+        return swigValues[i];
+    throw new IllegalArgumentException("No enum " + Result.class + " with value " + swigValue);
+  }
+
+  private Result(String swigName) {
+    this.swigName = swigName;
+    this.swigValue = swigNext++;
```
[the remark this answers is not in the document]
Comment (baierd) [reply lib/native/source/libbitwuzla/src/org/sosy_lab/java_smt/solvers/bitwuzla/api/Result.java:43]: Bitwuzla dies anyways for multithreaded use ;D

File: src/org/sosy_lab/java_smt/solvers/bitwuzla/BitwuzlaFormulaManager.java (hunk 7 above)
Comment (baierd) [reply src/org/sosy_lab/java_smt/solvers/bitwuzla/BitwuzlaFormulaManager.java:209]: No we don't want to do this, but it works for now. I will ask the devs to provide a method for us that takes a arbitrary SMTIB2 String and returns a Term.
If this does not work, i'll add FP to the independent SMTLIB2 layer and add it using that.

File: src/org/sosy_lab/java_smt/solvers/bitwuzla/BitwuzlaModel.java (hunk 8 above)
Comment (baierd) [reply src/org/sosy_lab/java_smt/solvers/bitwuzla/BitwuzlaModel.java:205]: I've added a test.
The results are:
```
  // -0 == 0 for FP equals is SAT
  // -0 != 0 for equals
  // nan != nan for FP equals
  // nan == nan for equals is SAT
  // pos inf != neg inf FP eq
  // pos inf != neg inf eq
  ```

File: src/org/sosy_lab/java_smt/solvers/bitwuzla/BitwuzlaQuantifiedFormulaManager.java (hunk 9 above)
Comment (baierd) [reply src/org/sosy_lab/java_smt/solvers/bitwuzla/BitwuzlaQuantifiedFormulaManager.java]: I pulled it into the loop and did not reuse the array.
I also found that the quantifier tests did block all solvers that don't support LIA and BV at the same time and fixed that.
I also added `requireQuantififerElimination()` (which might be a bit ugly) and changed the behavior of CVC5 as i looked it up and did not make sense.

File: src/org/sosy_lab/java_smt/solvers/bitwuzla/BitwuzlaSolverContext.java (hunk 10 above)
Comment (baierd) [reply src/org/sosy_lab/java_smt/solvers/bitwuzla/BitwuzlaSolverContext.java:158]: Aren't we setting all configs via reflections...? :D

File: src/org/sosy_lab/java_smt/solvers/opensmt/OpenSmtFormulaManager.java (hunk 11 above)
Comment (baierd) [reply src/org/sosy_lab/java_smt/solvers/opensmt/OpenSmtFormulaManager.java:86]: I just fixed bugs along the way for other solvers (CVC5 is now also affected). Do you want me to remove those and do them separately?

Comment (baierd) [reply src/org/sosy_lab/java_smt/solvers/opensmt/OpenSmtFormulaManager.java:86]: I've also opened a issue for the issue you pointed out!

File: build/build-publish-solvers/solver-bitwuzla.xml (hunk 6 above)
[the remark this answers is not in the document]
Comment (PhilippWendler) [reply build/build-publish-solvers/solver-bitwuzla.xml:198]: Hm, I wouldn't trust `ldd` to tell you details about what is in the files. `ldd` actually loads the file and then informs you about what happens (kinda). Pretty sure it also deduplicates transitive dependencies in the whole dependency graph.

For actually inspecting the binary, try using different tools. Maybe `objdump` can print this, or similar tools.

File: src/org/sosy_lab/java_smt/solvers/opensmt/OpenSmtFormulaManager.java (hunk 11 above)
Comment (baierd) [reply src/org/sosy_lab/java_smt/solvers/opensmt/OpenSmtFormulaManager.java]: I've now removed the changes to other solvers and extracted them into 2 branches, 1 for parsing and 1 for quantifiers

File: src/org/sosy_lab/java_smt/solvers/bitwuzla/BitwuzlaSolverContext.java (hunk 10 above)
Comment (baierd) [reply src/org/sosy_lab/java_smt/solvers/bitwuzla/BitwuzlaSolverContext.java:196]: Fixed.

File: lib/native/source/libbitwuzla/bitwuzla.i (hunk 14 above)
[the remark this answers is not in the document]
Comment (baierd) [reply lib/native/source/libbitwuzla/bitwuzla.i:303]: I've opened an issue for this so that we can provide the call for everyone in the future.

**Comment by kfriedberger** [comment]
Can someone give details on the compilation process of Bitwuzla:
- On which target system is it expected to be executed? Is Ubuntu 18.04 sufficient or is something newer required?
- Was the checked-in artifact tested on several systems? (Gitlab-Runner does not execute Bitwuzla. Ubuntu 22.04 does not execute Bitwuzla.)
- Was Bitwuzla compiled with an existing Docker-file for Ubuntu 18.04? Was it modified?
- Why was Bitwuzla not added to the main documentation in the Readme-file?
- Was Bitwuzla uploaded to all required artifact systems, namely Ivy and Maven?
